### PR TITLE
(2/4) MRF Primitive Visitor

### LINKED ISSUE #235: Create a visitor to create radio packets from Primitives
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
We should use a visitor to create radio packets from the `Primitive` classes to remove the responsibility from the `Primitive` classes.

**Describe the solution you'd like**
Create a new `RadioSerializerPrimitiveVisitor` that extends from the `PrimitiveVisitor` interface. This visitor should be responsible for creating a radio packet from the information in each Primitive. This includes packing the parameters into the parameters and bits, as well as the `Primitive Type` id (this will be handled by the visitor instead of #221 )

**Describe alternatives you've considered**
Leaving things as they are. This means that the `Primitive` classes are partially aware and in-control of the radio encoding, because they must return the parameters and "extra bits" in the correct order from the `getParameters` and `getExtraBits` functions. This is not ideal.

**Additional context**
A similar approach is being taken to use a Visitor to encode the `Primitives` to ROS messages. See #226

### LINKED ISSUE #334: Add unit tests to MRF primitive visitor
### Description of the task

Add unit tests to the MRF primitive visitor added in #333.

### Outline the task in more detail

Add unit tests to each of the primitive subclasses to be visited by the visitor; constructs a RadioPrimitive struct and checks that the members are correctly formatted.
### Acceptance criteria

- [ ] GTest added for MRFPrimitiveVisitor class
- [ ] Tests pass...?

### PR DESCRIPTION
**Pull request by henryymliu**
### Description
*Give a high-level description of the changes in this PR*
Adds a visitor that will convert primitives into a format that is easier for conversion into radio packets.

### Testing Done
Field tested most of the primitives. Unit testing TBD in #334 

### Resolved Issues
resolves #235 
resolves #334 

### Length Justification

N/A

### Review Checklist
*(Please check every item to indicate your code complies with it (by changing `[ ]`->`[x]`). This will hopefully save both you and the reviewer(s) a lot of time!)*
***It is the reviewers responsibility to also make sure every item here has been covered***
- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom` 
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

***Feel free to make additions of things that we should be checking to this file if you think there's something missing!!!!***

### CODE REVIEW

File: src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    r_prim.param_array = {catch_primitive.getVelocity(),
+                          catch_primitive.getDribblerSpeed(),
+                          catch_primitive.getMargin()};
+    r_prim.extra_bits  = 0;
+}
+
+void MRFPrimitiveVisitor::visit(const ChipPrimitive &chip_primitive)
+{
+    r_prim.prim_type   = PrimitiveType::SHOOT;
+    r_prim.param_array = {chip_primitive.getChipOrigin().x() * 1000,
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.cpp]: What is `1000` here? Could we make it a descriptively named variable instead? Same goes for the all the "magic numbers" here.

File: src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.h [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+     */
+    void visit(const StopPrimitive &stop_primitive) override;
+
+    /**
+     * Returns the most recent serialized packet created by this
+     * MRFPrimitiveVisitor. This is the radio packet created by
+     * one of the 'visit' functions
+     *
+     * Calling this function before this Visitor has been called at least once has
+     * undefined behavior
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.h]: Could we make it return an `std::optional` instead? I'd rather we avoid "undefined behavior" when at all possible, as it makes things much harder to debug.

Alternatively, we could just have it throw an exception.

File: src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.h [hunk 3]
```diff
@@ -0,0 +1,112 @@
+#pragma once
+
+#include "ai/primitive/visitor/primitive_visitor.h"
+#include "shared/primitive_type.h"
+
+/**
+ * This struct stores the components of a translated primitive to be sent over radio.
+ */
+typedef struct _RadioPrimitive
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.h]: Just for consistency, could we rename this to:
``` suggestion
typedef struct RadioPrimitive_t
```

File: src/thunderbots/shared/primitive_type.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+/**
+ * This enum represents the IDs of the primitives implemented in firmware.
+ */
+enum class PrimitiveType
```
Comment (garethellis0) [inline src/thunderbots/shared/primitive_type.h]: Could we rename to `FirmwarePrimitiveType` just to make it crystal-clear that this is firmware specific?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Unless there are special circumstances here, I don't see any reason for pushing unit testing off into another issue? Unit tests should be merged along with the code.

File: src/thunderbots/shared/primitive_type.h (hunk 4 above)
Comment (MathewMacDougall) [reply src/thunderbots/shared/primitive_type.h]: Agree with this

File: src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.cpp]: I expect this is a conversion to millimeters. We have descriptive named constants to do conversions like these in `shared/constants.h`

**Comment by henryymliu** [comment]
I had no time to work on this for the past week, which is why I wanted to push testing to another ticket.

re: throwing exception/std::optional for `visit`, I will begrudgingly do it but would much rather have it done in a separate ticket because a) I just want this stuff merged in and b) the grsim visitor also has this behaviour

**Comment by garethellis0** [comment]
Hmm, I definitely support wanting to get stuff done fast, but I still don't see a reason to merge it without testing if it's not blocking anything?

**Comment by henryymliu** [comment]
this has been field tested, and I want to move on to work on other things. Call me lazy, but I really don't want to work on this anymore.

File: src/thunderbots/shared/constants.h [hunk 5]
```diff
@@ -37,3 +38,6 @@ const double MILLISECONDS_PER_SECOND      = 1000.0;
 const double SECONDS_PER_MICROSECOND      = 1.0 / 1000000.0;
 const double SECONDS_PER_MILLISECOND      = 1.0 / 1000.0;
 const double MILLISECONDS_PER_MICROSECOND = 1.0 / 1000.0;
+
+// Converts dribbler RPM to a smaller number firmware uses
+const double DRIBBLER_RPM_TO_RADIO_CONVERSION_FACTOR = 1.0 / 300.0;
```
Comment (garethellis0) [inline src/thunderbots/shared/constants.h]: github complaining about end-of-file

File: src/thunderbots/software/CMakeLists.txt [hunk 6]
```diff
@@ -642,7 +642,21 @@ if (CATKIN_ENABLE_TESTING)
...
+            ai/primitive/direct_velocity_primitive.cpp
+            ai/primitive/kick_primitive.cpp
+            ai/primitive/pivot_primitive.cpp
+            ai/primitive/move_primitive.cpp
+            ai/primitive/movespin_primitive.cpp
+            ai/primitive/stop_primitive.cpp
+            ai/primitive/directwheels_primitive.cpp
+            ai/primitive/primitive.cpp
+            )
+    target_link_libraries(mrf_primitive_visitor_test ${catkin_LIBRARIES}
```
Comment (garethellis0) [inline src/thunderbots/software/CMakeLists.txt:713]: Looks like we've accidentally nuked the `target_link_libraries` for `grsim_catch_primitive_test` here, which I think is what's causing the failures?

File: src/thunderbots/shared/constants.h (hunk 5 above)
Comment (henryymliu) [reply src/thunderbots/shared/constants.h]: fixed

File: src/thunderbots/software/CMakeLists.txt (hunk 6 above)
Comment (henryymliu) [reply src/thunderbots/software/CMakeLists.txt:713]: Ah that would do it, fixed.

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

**Comment by garethellis0** [comment]
Also, this PR should now resolve #334 as well?

File: src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.cpp [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+void MRFPrimitiveVisitor::visit(const DribblePrimitive &dribble_primitive)
+{
+    radio_prim              = RadioPrimitive();
+    radio_prim->prim_type   = FirmwarePrimitiveType::DRIBBLE;
+    radio_prim->param_array = {
+        dribble_primitive.getDestination().x() * MILLIMETERS_PER_METER,
+        dribble_primitive.getDestination().y() * MILLIMETERS_PER_METER,
+        dribble_primitive.getFinalAngle().toRadians() * CENTIRADIANS_PER_RADIAN,
+        // There is no division by 300 for RPM in this primitive
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.cpp]: wut. Just out of curiosity, why? also, as a quick suggestion so we don't have values duplicated:
``` suggestion
        // For this primitive, we don't divide the RPM
```

**Review by garethellis0 — DISMISSED** [review]
LGTM; just one comment/question above

File: src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.cpp (hunk 7 above)
Comment (henryymliu) [reply src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.cpp]: I don't know the answer to this question... was just reading off the firmware implementation of the primitive

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

File: src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.cpp (hunk 7 above)
Comment (garethellis0) [reply src/thunderbots/software/radio_communication/visitor/mrf_primitive_visitor.cpp]: All good, was just curious

**Pull request merged**